### PR TITLE
Suggestion 4: Add Fault Tests 

### PR DESCRIPTION
**Pull request by trimixlover**
# Description

Quoting auditors report:
> - f validators were not started;  
> - f validators are up but not connected to the network; 
> - f validators crash simultaneously and then recover; 
> - f validators crash at different times and then recover one by one at random times;
> - f validators crash and recover after a significant delay (e.g., after 10 minutes); and
> - f validators drop every second message (message loss is 50%).

1. Validators not started are implicitly tested with a `TestDropMaxFaulty.` We don't discriminate between hights of the sequence ran because it is not in the scope of go-ibft to sync missing blocks.
2. This test is very relevant when testing by using orchestrated environment. In drop tests there is no difference between _not started_ and _not connected_.
3. and 4. are dealt with implementing number `TestAllFailAndGraduallyRecover` as it is a more difficult scenario for the network to deal with.
5. This kind of test will heavily impact our CI/CD process and should definitely be implemented in separate effort. Maybe where this tests are written to run on actually test cluster.
6. This test is covered with `TestMaxFaultyPlusOneDroppingMessages`

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have added sufficient documentation in code

### CODE REVIEW

File: core/drop_test.go [hunk 1]
```diff
@@ -1,19 +1,222 @@
...
+	err := cluster.progressToHeight(5*time.Second, 1)
+	assert.NoError(t, err, "unable to reach height: %w", err)
+	assert.Equal(t, uint64(1), cluster.latestHeight)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+
+	insertedBlocks = make([][]byte, numNodes) // Purge
+
+	// Stop all nodes and make sure no blocks are written
+	cluster.stopN(len(cluster.nodes))
+	cluster.progressToHeight(2*time.Second, 2)
```
Comment (Kourin1996) [inline core/drop_test.go:72]: timeout must be same as the one used for others cause blocks are not filled due to lack of time.

File: core/drop_test.go [hunk 2]
```diff
@@ -1,19 +1,222 @@
...
+	assertNoInsertedBlocks(insertedBlocks, t)
+
+	// Start all and expect valid blocks to be written again
+	cluster.startN(len(cluster.nodes))
+	cluster.progressToHeight(5*time.Second, 10)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+}
+
+func assertNoInsertedBlocks(blocks [][]byte, t *testing.T) {
```
Comment (Kourin1996) [inline core/drop_test.go:81]: Nitpick, t comes first generally

File: core/drop_test.go [hunk 3]
```diff
@@ -1,19 +1,222 @@
...
+	t.Helper()
+
+	assertNInsertedBlocks(0, blocks, t)
+}
+
+func assertNInsertedBlocks(n int, blocks [][]byte, t *testing.T) {
```
Comment (Kourin1996) [inline core/drop_test.go:87]: What calls `assertNInsertedBlocks` except for `assertNoInsertedBlocks`? You can check in ` assertNoInsertedBlocks` directly.

File: core/drop_test.go [hunk 4]
```diff
@@ -1,19 +1,222 @@
...
+	for _, block := range blocks {
+		if !bytes.Equal(block, nil) {
+			writtenBlocks++
+		}
+	}
+
+	assert.True(t, n == writtenBlocks)
+}
+
+func assertValidInsertedBlocks(blocks [][]byte, t *testing.T) {
```
Comment (Kourin1996) [inline core/drop_test.go:100]: I think it's better to give an expected value as an argument

File: core/drop_test.go [hunk 5]
```diff
@@ -1,19 +1,222 @@
...
+						}
+
+						c.gossip(message)
+					}},
+				)
+			}
+		},
+	)
+
+	cluster.makeNFaulty(int(cluster.maxFaulty()))
```
Comment (Kourin1996) [inline core/drop_test.go:148]: Is this MaxFaultyPlusOne?

**Review by Kourin1996 — APPROVED** [review]

**Review by b00ris — APPROVED** [review]

File: core/drop_test.go (hunk 1 above)
Comment (trimixlover) [reply core/drop_test.go]: Done

File: core/drop_test.go (hunk 2 above)
Comment (trimixlover) [reply core/drop_test.go]: Fixed

File: core/drop_test.go (hunk 3 above)
Comment (trimixlover) [reply core/drop_test.go]: There are some cases

File: core/drop_test.go (hunk 5 above)
Comment (trimixlover) [reply core/drop_test.go:142]: Yes i renamed the test since the max faulty plus one can actually pass as a test

File: core/drop_test.go (hunk 4 above)
Comment (trimixlover) [reply core/drop_test.go]: Used for just one test, would increase the complexity of the method enough that it should be implemented within the test it self.

**Pull request merged**